### PR TITLE
evmrs: add feature "stack-array"

### PR DESCRIPTION
**Pull request by LorenzSchueler**
This PR adds the feature "stack-array". When this feature is enabled the Stack holds a [MaybeUninit<u256>; 1024] instead of Vec<u256>.

### CODE REVIEW

File: rust/src/types/stack.rs [hunk 1]
```diff
@@ -1,19 +1,112 @@
...
+                &mut self.data[..self.len],
+            )
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    pub unsafe fn push(&mut self, value: u256) {
```
Comment (LuisPH3) [inline rust/src/types/stack.rs:54]: I think you can limit the use of unsafe to this object, and do the range check, like you do with pop.

File: rust/src/types/stack.rs [hunk 2]
```diff
@@ -1,19 +1,112 @@
...
+        *self.data.get_unchecked_mut(self.len) = MaybeUninit::new(value);
+        self.len += 1;
+    }
+
+    pub fn pop_array<const N: usize>(&mut self) -> Option<[u256; N]> {
+        if self.len < N {
```
Comment (LuisPH3) [inline rust/src/types/stack.rs:60]: dead code?  check overflow should prevent this.

File: rust/src/types/stack.rs [hunk 3]
```diff
@@ -1,19 +1,112 @@
...
+    pub fn as_slice(&self) -> &[u256] {
+        self.0.as_slice()
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
     }
 
     pub fn push(&mut self, value: impl Into<u256>) -> Result<(), FailStatus> {
         self.check_overflow_on_push()?;
```
Comment (LuisPH3) [inline rust/src/types/stack.rs:102]: if we move the overflow check inside of push (like pop does) 
we can do the overflow check only for the vector case.  and remove the unsafe block from this scope.

File: rust/src/types/stack.rs [hunk 4]
```diff
@@ -22,23 +115,30 @@ impl Stack {
...
         self.check_underflow(N)?;
 
-        let new_len = self.0.len() - N;
-        let mut array = [u256::ZERO; N];
-        array.copy_from_slice(&self.0[new_len..]);
-        self.0.truncate(new_len);
-        Ok(array)
+        #[cfg(feature = "stack-array")]
+        {
+            Ok(self.0.pop_array().unwrap())
```
Comment (LuisPH3) [inline rust/src/types/stack.rs:127]: As mentioned before, there is a duplicated bound check here.
Maybe is easier to write: `return self.0.pop_array()` return a result and do the bounds check in the alternative implementation. 
same for Push

**Review by LuisPH3 — CHANGES_REQUESTED** [review]
This is a good example of low level mem management in rust. I would like to reduce a little more the scope of the unsafe regions, tough.

File: rust/src/types/stack.rs (hunk 1 above)
Comment (LorenzSchueler) [reply rust/src/types/stack.rs:54]: Yes but without feature "stack-array" I have to do the range check in Stack::pop. When the feature is enabled I could do it in StackRaw::pop but that would mean that it is done in two different places depending on the feature.

File: rust/src/types/stack.rs (hunk 2 above)
Comment (LorenzSchueler) [reply rust/src/types/stack.rs:60]: Yes. I think pop_array should be unsafe and assume that the check was already done and then return [u256; N] directly.

File: rust/src/types/stack.rs (hunk 3 above)
Comment (LorenzSchueler) [reply rust/src/types/stack.rs:102]: See earlier comment. I don't like the idea of doing the check in different places.

**Comment by LorenzSchueler** [comment]
One other idea: what do you think about renaming StackRaw to Stack and using it directly instead of wrapped in a Stack. That would solve all of those ^ issues.

**Review by LuisPH3 — APPROVED** [review]

**Pull request merged**